File: DSA/Strings/JS/GroupAnagram.js
/* 
Given an array of strings consisting of lower case English letters, group the anagrams together. You can 
return the answer in any order. An anagram is a word or phrase formed by rearranging the letters of a 
different word or phrase, using all the original letters exactly once.

*/

const groupingAnagrams = function (arr) {
  const hashedArr = {};
  const sortedArr = arr.map((char) => char.split('').sort().join(''));
  for (let i = 0; i < arr.length; i++) {
    let sortedWord = sortedArr[i];
    if (hashedArr[sortedWord]) hashedArr[sortedWord].push(arr[i]);
    else hashedArr[sortedWord] = [arr[i]];
  }

  return Object.values(hashedArr);
};

const example = ['arc', 'abc', 'car', 'cat', 'act', 'acb', 'atc'];
const example1 = ['eat', 'tea', 'tan', 'ate', 'nat', 'bat'];
console.log(groupingAnagrams(example));
console.log(groupingAnagrams(example1));

File: DSA/Strings/py/GroupAnagram.py
'''
Given an array of strings consisting of lower case English letters, group the anagrams together. You can 
return the answer in any order. An anagram is a word or phrase formed by rearranging the letters of a 
different word or phrase, using all the original letters exactly once.
'''

def grouping_anagram(list):
  sorted_list = [''.join(sorted(word)) for word in list]
  anagram_dict = {}

  for i in range(len(sorted_list)):
    if sorted_list[i] in anagram_dict:
      anagram_dict[sorted_list[i]].append(list[i])
    else:
      anagram_dict[sorted_list[i]] = [list[i]]
  
  return anagram_dict.values()

print(grouping_anagram(['arc', 'abc', 'car', 'cat', 'act', 'acb', 'atc']))
print(grouping_anagram(['eat','tea', 'tan', 'ate', 'nat', 'bat']))